src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -407,6 +407,8 @@ inferBinder val (NamedBinder name binder) = do
   return $ M.insert name val m
 inferBinder val (PositionedBinder pos _ binder) =
   warnAndRethrowWithPosition pos $ inferBinder val binder
+-- | NOTE: Fail!
+inferBinder val (TypedBinder ty binder) = val =?= ty >> inferBinder val binder
```

Comment:
This is basically correct, but we might have to be careful about polymorphic types. In general, you'll need to use `subsumes`, not unification here, but the subsumption check can change the expression by elaborating it, so we'd need to take care of that too.

The alternative is to just restrict typed binders to monotypes, which isn't a bad option for now. In that case, you could keep the unification here, and return `True` in `binderRequiresMonotype`.

I guess it depends what we want typed binders for. If we think of them as for guiding things like type class search, then a monotype restriction is probably fine, but if we want to use them to switch into checking mode for things like rank-N types, then we might want to implement the full subsumption check.